src/ILLink.CodeFix/DAMCodeFixProvider.cs

```diff
@@ -63,55 +96,31 @@ public override async Task RegisterCodeFixesAsync (CodeFixContext context)
 			if (await document.GetSyntaxRootAsync (context.CancellationToken).ConfigureAwait (false) is not { } root)
 				return;
 			var diagnostic = context.Diagnostics[0];
-			SyntaxNode diagnosticNode = root.FindNode (diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
-			if (await document.GetSemanticModelAsync (context.CancellationToken).ConfigureAwait (false) is not { } model)
+			if (diagnostic.AdditionalLocations.Count == 0)
 				return;
-			// Note: We get the target symbol from the diagnostic location. 
-			// This works when the diagnostic location is a method call, because the target symbol will be the called method with annotations, but won't work in general for other kinds of diagnostics.
-			if (model.GetSymbolInfo (diagnosticNode).Symbol is not { } targetSymbol)
+			if (root.FindNode (diagnostic.AdditionalLocations[0].SourceSpan, getInnermostNodeForTie: true) is not SyntaxNode attributableNode)
 				return;
-			if (model.Compilation.GetTypeByMetadataName (FullyQualifiedAttributeName) is not { } attributeSymbol)
+			if (diagnostic.Properties["attributeArgument"] is not string stringArgs || stringArgs.Contains (","))
 				return;
 
-			if (diagnosticNode is not InvocationExpressionSyntax invocationExpression)
-				return;
-
-			var arguments = invocationExpression.ArgumentList.Arguments;
-
-			if (arguments.Count > 1)
-				return;
-
-			if (arguments.Count == 1) {
-				if (arguments[0].Expression is not LiteralExpressionSyntax literalSyntax
-					|| literalSyntax.Kind () is not SyntaxKind.StringLiteralExpression) {
-					return;
-				}
-			}
-
-			// N.B. May be null for FieldDeclaration, since field declarations can declare multiple variables
-			var attributableSymbol = (invocationExpression.Expression is MemberAccessExpressionSyntax simpleMember
-					&& simpleMember.Expression is IdentifierNameSyntax name) ? model.GetSymbolInfo (name).Symbol : null;
-
-
-			if (attributableSymbol is null)
+			var syntaxGenerator = SyntaxGenerator.GetGenerator (document);
+			var attributeArguments = new[] { syntaxGenerator.AttributeArgument (syntaxGenerator.MemberAccessExpression (syntaxGenerator.DottedName ("System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes"), stringArgs)) };
+			if (await document.GetSemanticModelAsync (context.CancellationToken).ConfigureAwait (false) is not { } model)
```

Comment:
Do you mean to shuttle the lines that reference the model into `AddAttributeAsync()`? That seems like a straightforward fix

Reply:
It can be very straightforward, but we also try to avoid offering no-op code fixes (a code fix that doesn't actually do anything), so sometimes just moving the check doesn't matter. Since the no-op case here looks to be a scenario that should never happen for a user, it's fine to move it into the callback as you were thinking.